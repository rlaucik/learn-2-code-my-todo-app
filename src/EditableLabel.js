import React from 'react';

export class EditableLabel extends React.Component {
    constructor(props) {
        super(props);

        this.state = {
            isEditing: false,
            text: props.text,
        }
        this.inputRef = React.createRef();
    }

    componentDidMount() {
        document.addEventListener('mousedown', this.handleClickOutside);
    }

    componentWillUnmount() {
        document.removeEventListener('mousedown', this.handleClickOutside);
    }

    render() {
        if (this.state.isEditing) {
            return(
                <input
                    type="text"
                    ref={this.inputRef}
                    value={this.state.text}
                    onChange={this.onChange}
                    onKeyPress={this.onKeyPress}
                />
            );
        }
        return (
            <label
                onDoubleClick={this.onDoubleClick}
            >
                {this.props.text}
            </label>
        );
    }

    onDoubleClick = () => {
        this.setState({
            isEditing: true,
        }, () => {
            this.inputRef.current.focus();
        });
    }

    onChange = event => {
        this.setState({
            text: event.target.value,
        });
    }

    handleSave = () => {
        const value = this.inputRef.current.value;
        this.setState({
            isEditing: false,
        });
        this.props.onEdit(value, this.props.itemIndex);
    }

    onKeyPress = event => {
        if (event.key === 'Enter') {
            this.handleSave();
        }
    }

    handleClickOutside = (event) => {
        if (this.inputRef.current && !this.inputRef.current.contains(event.target)) {
            this.handleSave();
        }
    }
}
