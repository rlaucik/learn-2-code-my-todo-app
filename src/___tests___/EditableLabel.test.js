import React from 'react';
import { shallow } from 'enzyme';
import { EditableLabel } from '../EditableLabel';

describe('EditableLabel', () => {
    it('renders label when not editing', () => {
        const component = shallow(
            <EditableLabel
                text="some text"
            />
        );
        
        expect(component).toMatchSnapshot();
    });
    it('renders input when in editing mode', () => {
        const component = shallow(
            <EditableLabel
                text="text to edit"
            />
        );

        component.setState({ isEditing: true });
        
        expect(component).toMatchSnapshot();
    });
});