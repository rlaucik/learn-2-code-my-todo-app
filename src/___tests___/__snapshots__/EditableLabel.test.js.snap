// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditableLabel renders input when in editing mode 1`] = `
<input
  onChange={[Function]}
  onKeyPress={[Function]}
  type="text"
  value="text to edit"
/>
`;

exports[`EditableLabel renders label when not editing 1`] = `
<label
  onDoubleClick={[Function]}
>
  some text
</label>
`;
